Add support for messaging between different browsing contexts

jgraham opened the pull request:
This will be split up to land, and will have one or more associated RFCs.

The features added are:
* Support for giving contexts known ids through a `uuid` parameter in the URL.
* Support for executing script in remote contexts using testdriver, and an implementation for wptrunner based on WebDriver.
* Support for cross-context messaging based on send/poll/recv methods in testdriver, using the server stash as a backend.

jgraham committed:
Add support for shared queues in the stash

This allows stash-owned queues that can then be accessed across
multiple servers etc. In particular it can be used as the backend for
a shared message queue in which multiple browsing contexts can send
messages using the stash.

jgraham committed:
Try to set the stack property on the status for cleanup errors

This makes them easier to diagnose.

jgraham committed:
Implementation of websocket-backed cross-context messaging

This allows creating message channels that can cross multiple browsing
contexts or script execution environments.

A simple channel pair can be created using the channel function e.g.

let [recvChan, sendChan] = await channel();

This generates a channel pair such that messages sent on sendChan will
arrive on recvChan.

Channels are multiple producer, single consumer, so there can only be
a single ReadChannel for each network, but multiple SendChannels.

On top of this underlying abstraction there's a higher level API which
provides known message structures. There is currently a command
message type where each command has an id, a method, a set of params,
and a respChannel which is a channel to use to reply to the command.

Currently two commands are provided: call, which runs a function in a
remote context, and postMessage, which sends a message to a remote
context. Messages and script parameters are serialized using a
WebDriver BiDi like remote value serialization algorithm which allows
passing complex objects over the wire.

To work with multiple contexts, one takes the following actions:

* In the context that will recieve messages, use:

  let recvChannel = await start_global_channel();

  This starts listening for messages on a channel defined by the
  testdriver context id (typically in the URL; maybe this should just
  be always in the URL which could also remove the testdriver.js
  dependency).

  If only executeScript is being used nothing more needs to be
  done. For postMessage, one can register message handlers using
  `recvChannel.addMessageHandler("message", fn)`.

* In the context that will send messages, use the uuid of the remote
  context to create a RemoteGlobal object. This will allow sending
  messages to the remote.

  let remote = RemoteGlobal(uuid);
  let result = remote.call(selector =>
    document.querySelector(selector).textContent,
    "body")

  remote.postMessage("done");

Message channels are backed with websockets, which use the stash as
the data storage layer. For use cases where it's important to close
all websockets (e.g. testing bfcache), the closeAllChannelSockets
helper is provided.

hiroshige-g commented on html/browsers/browsing-the-web/back-forward-cache/resources/executor.html:
```diff
@@ -0,0 +1,117 @@
+// This ensures that no subsequent commands will be buffered in the underlying
+// websocket before the navigation starts. This has to be done in the caller
+// to avoid a race condition between the caller sending a new command and the
+// socket being closed.
```
In https://github.com/web-platform-tests/wpt/pull/28950,

- I have something similar to `closeAllChannelSockets()` that suspends polling network requests on the remote context side,
- but no `pause()`-ish things, relying on that the caller doesn't send any new commands until the page becomes inactive

so I'm curious how `pause()` works.
Particularly, are both of `pause()` and `closeAllChannelSockets()` needed?
Is there a race condition between these two?

hiroshige-g commented on html/browsers/browsing-the-web/back-forward-cache/resources/executor.html:
```diff
@@ -0,0 +1,117 @@
+  },
+  {once: true});
+
+async function navigateTo(url) {
```
As there are ways to trigger navigation other than `location.href` (e.g. form submission), I'd keep `prepateNavigation()` and `execute_script()` for navigation purpose.

foolip commented on infrastructure/channels/serialize-data.js:
```diff
@@ -0,0 +1,38 @@
+    "false": {input: false},
+    "bigint": {input: 1n},
+    "RegExp": {input: /abc/g},
+    "Date": {input: new Date('December 17, 1995 03:24:00')},
```
Someone's birthday? :)

jgraham replied:
Presumably lots of people's, but I don't (think I) know any of them (I think I just copied it from somewhere).

foolip commented on infrastructure/channels/serialize-data.js:
```diff
@@ -0,0 +1,38 @@
+
+let cyclicArray = [2];
```
This would seem more consistent with the theme that follows:

```suggestion
let cyclicArray = [1];
```

foolip commented on infrastructure/channels/serialize_child.html:
```diff
@@ -0,0 +1,167 @@
+
+let lastData;
+
+// Hack: these will be recreated as assert failures in the caller
```
Why though? Can't we fail in any manner and get the same result? Does this depend on how testharness.js treats "AssertionError" differently in some place?

foolip commented on infrastructure/channels/serialize_child.html:
```diff
@@ -0,0 +1,167 @@
+
+function compareResult(name, actual) {
+    let obj = objects[name]
+    let expected = obj.hasOwnProperty("output") ? obj.output : obj.input;
```
Why do we need to ensure that `output` is an own property, but don't care for `input`?

foolip commented on infrastructure/channels/serialize_child.html:
```diff
@@ -0,0 +1,167 @@
+    }
+    if (["undefined", "string", "boolean", "number", "bigint"].includes(typeof expectedValue) ||
+        actualValue === null) {
+        if (Number.isNaN(expectedValue)) {
```
Could we use `Object.is` to get NaN and signed zero handling for free?

foolip commented on infrastructure/channels/serialize_child.html:
```diff
@@ -0,0 +1,167 @@
+    if (expectedValue.constructor && actualValue.constructor && expectedValue.constructor.name !== actualValue.constructor.name) {
+        throw new Error(`Constructors differ, expected ${expectedValue.constructor.name}, got ${actualValue.constructor.name}`);
+    }
+    if (expectedValue.constructor.name === "SendChannel") {
```
Going just by the local code, `expectedValue.constructor` could still be undefined here. Otherwise, why bother checking it above?

foolip commented on infrastructure/channels/serialize_child.html:
```diff
@@ -0,0 +1,167 @@
+    else if (expectedValue.constructor.name === "RegExp") {
+        if (expectedValue.source !== actualValue.source ||
+            expectedValue.flags !== actualValue.flags) {
+            throw new Error(`Regexps differ, expected ${expectedValue}, got ${actualValue}`);
```
```suggestion
            throw new Error(`RegExps differ, expected ${expectedValue}, got ${actualValue}`);
```

foolip commented on infrastructure/channels/serialize_child.html:
```diff
@@ -0,0 +1,167 @@
+            throw new Error(`Regexps differ, expected ${expectedValue}, got ${actualValue}`);
+        }
+    } else if (expectedValue.constructor.name == "Date") {
+        if (expectedValue.toDateString() !== actualValue.toDateString()) {
```
Using `valueOf()` should be faster here, and avoid any situation where `toDateString()` returns the same string despite the underlying value not being exactly the same.

It could make the error message harder to understand, though, so I'm not 100% sure of this suggestion. You pick :)

jgraham committed:
Allow all resources to use setTimeout

These aren't tests, so it can be assumed we're using it carefully.

foolip commented on infrastructure/channels/serialize_child.html:
```diff
@@ -0,0 +1,167 @@
+            throw new Error(`Errors differ, expected ${expectedValue}, got ${actualValue}`);
+        }
+    } else if (Array.isArray(expectedValue)) {
+        seenActual = seen.has(actualValue);
```
`let seenActual` is missing somewhere for this to not leak into the global scope. Not here though, but above?

foolip commented on infrastructure/channels/serialize_child.html:
```diff
@@ -0,0 +1,167 @@
+        if (actualValue.size !== expectedValue.size) {
+            throw new Error(`Set sizes differ, expected ${expectedValue.size}, got ${actualValue.size}`);
+        }
+        // So for a set we can't really check if the values are the same
```
You mean because we can't sort them? I think we could check if they're the same, but it would be O(n^2), and with nested sets worse still.

foolip commented on infrastructure/channels/serialize_child.html:
```diff
@@ -0,0 +1,167 @@
+
+
+        // Compare as a general Object
+        if (Object.keys(actualValue).length !== Object.keys(expectedValue).length) {
```
Calling both `Object.keys(expectedValue)` here and `Object.entries(expectedValue)` is a bit wasteful. Now, if one wanted to optimize this one should probably avoid creating those temporary arrays entirely, but let's do the simple optimization at least.

foolip reviewed (commented):
This is a lot! I've reviewed up to infrastructure/channels/serialize_child.html...

jgraham committed:
Make the stash manager a class property

The current setup assumed that we had one Stash instance per process
and so if the lock property on the class was initialised then we must
also have initialised the manager. The way that websockets handlers
are loaded breaks this assumption, so we need to ensure that the
lifecycle of the lock and of the manager object are identical. Sharing
the manager state between instances makes sense since we're already
sharing the proxy we get from the stash.

jgraham commented on html/browsers/browsing-the-web/back-forward-cache/resources/executor.html:
```diff
@@ -30,23 +31,31 @@
-
-const executor = new Executor(uuid);
+  window.addEventListener('pageshow',
+                          async () => {
```
I didn't fix this because I don't intend to land the test changes as part of this PR; they're mostly here to verify how it works in practice, so where it's more complex it doesn't seem worth spending too much time on at this stage.

foolip replied:
Aha, I'll ignore test changes then!

foolip commented on resources/channel.sub.js:
```diff
@@ -0,0 +1,1003 @@
+        return "object";
+    }
+
+    let remoteObjectsById = new Map();
```
We can't use `WeakMap` because we need to look by UUID, but won't this map keep objects alive? Can it be avoided?

foolip commented on resources/channel.sub.js:
```diff
@@ -0,0 +1,1003 @@
+         * ``null``.
+         */
+        delete() {
+            if (remoteObjectsById.has(this.objectId)) {
+                remoteObjectsById.delete(this.objectId);
+            }
```
Simpler:

```suggestion
            remoteObjectsById.delete(this.objectId);
```

foolip commented on resources/channel.sub.js:
```diff
@@ -0,0 +1,1003 @@
+                    break;
+                case "regexp":
+                    value = {
+                        pattern:item.source,
```
```suggestion
                        pattern: item.source,
```

foolip commented on resources/channel.sub.js:
```diff
@@ -0,0 +1,1003 @@
+    self.RemoteObject = RemoteObject;
+
+    function serialize(obj) {
+        const stack = [{item: obj}];
```
We use `stack.push(...)` and stack.shift()`, so it's a queue, not a stack:

```suggestion
        const queue = [{item: obj}];
```

A comment describing what the items in the queue are and what the possible properties are would be really helpful too. I think that if I hadn't seen traversal-by-stack (or queue) a lot of times and hadn't reviewed the WebDriver BiDi serialization format, I'd be 100% lost here.

foolip commented on resources/channel.sub.js:
```diff
@@ -0,0 +1,1003 @@
+                    value = [];
+                    newTarget = {type: "map", value};
+                    for (let [targetKey, child] of item.entries()) {
+                        stack.push({item: targetKey, target: newTarget, targetKey: "key"});
```
The use of `targetKey` as a property here makes for confusion elsewhere I think. Neither "key" nor "values" are keys of anything, and it's also unrelated to the `targetKey` variable here.

foolip commented on resources/channel.sub.js:
```diff
@@ -0,0 +1,1003 @@
+                    target.value[targetKey] = result;
+                    break;
+                case "map":
+                    if (targetKey === "key") {
+                        target.value.push([]);
+                    }
+                    target.value[target.value.length - 1].push(result);
```
This needs a comment, it took me a lot of time to unpack. It depend on the traversal order given by the queue, where we have to get "key" first, and the last line here is used for both the "key" and "value" case. So we do end up pushing to the new `[]` twice, in the right order.

foolip commented on resources/channel.sub.js:
```diff
@@ -0,0 +1,1003 @@
+
+    function deserialize(obj) {
+        let deserialized = null;
+        let stack = [{item: obj, target: null}];
```
```suggestion
        let queue = [{item: obj, target: null}];
```

foolip commented on resources/channel.sub.js:
```diff
@@ -0,0 +1,1003 @@
+                        typeof self[item.value.type] === "function") {
+                        result = new self[item.value.type](item.value.message);
+                    } else {
+                        result = new Error(item.value.message);
```
If `item.value.type` is "Error", the above case should be taken. Can you add a comment that this isn't about deserializing `Error` objects, but non-`Error` values?

foolip commented on resources/channel.sub.js:
```diff
@@ -0,0 +1,1003 @@
+
+            if (target === null) {
+                if (deserialized !== null) {
+                    throw new Error("Tried to create multiple output values");
```
This is the same error as above, is it accurate here though? What's happened is we we've deserialized a value but there's nowhere to put it.

foolip commented on resources/channel.sub.js:
```diff
@@ -0,0 +1,1003 @@
+                    break;
+                case "map":
+                    if (targetKey === "key") {
+                        target.key = result;
```
Stuffing a value on `target` here makes me wonder. Like with serialization, it depends on the order that the queue is managed. We might just as well use a local variable, making it more clear that we only store one key at a time, and that things would blow up the order were different.

foolip replied:
Oh yeah, you're right!

foolip reviewed (commented):
I have reviewed infrastructure/ and resources/, only took me 2+ hours.

foolip commented on websockets/handlers/msg_channel_wsh.py:
```diff
@@ -0,0 +1,164 @@
+            queue, has_reader, writer_count = value
+            if direction == "read":
+                if has_reader:
+                    raise ValueError("Tried to start multiple readers for the same queue")
```
Shouldn't `has_reader` be set to True after this? Or what sets it to True in the stash?

jgraham replied:
Good catch.

foolip commented on websockets/handlers/msg_channel_wsh.py:
```diff
@@ -0,0 +1,164 @@
+    uuid, direction = parse_request(request)
+    log(uuid, f"Got web_socket_passive_closing_handshake {direction}")
+
+    if direction == "read":
```
A comment about what this cleanup is doing would be helpful.

foolip commented on websockets/handlers/msg_channel_wsh.py:
```diff
@@ -0,0 +1,164 @@
+    closed = False
+    while not closed:
+        try:
+            line = request.ws_stream.receive_message()
```
Is it really a line?

```suggestion
            msg = request.ws_stream.receive_message()
```

foolip commented on websockets/handlers/msg_channel_wsh.py:
```diff
@@ -0,0 +1,164 @@
+                closed = True
+                log(uuid, "Got client initiated close")
+            else:
+                logger.warning("Unexpected message on read socket  %s", line)
```
```suggestion
                logger.warning("Unexpected message on read socket %s", line)
```

foolip commented on websockets/handlers/msg_channel_wsh.py:
```diff
@@ -0,0 +1,164 @@
+
+def run_write(request, uuid, queue):
+    while True:
+        line = request.ws_stream.receive_message()
```
```suggestion
        msg = request.ws_stream.receive_message()
```

foolip commented on websockets/handlers/msg_channel_wsh.py:
```diff
@@ -0,0 +1,164 @@
+            log(uuid, f"Message queue already deleted")
+            return
+        if direction is None:
+            log(uuid, f"Force deleting message queue")
```
Where is it force deleted? We just return.

foolip replied:
Comment added now, thanks!

foolip reviewed (commented):
OK, I've reviewed the Python parts (tools/ and websockets/) too now.

jgraham committed:
Add documentation for the channels support

foolip commented on resources/channel.sub.js:
```diff
@@ -31,7 +31,20 @@
+     *
+     * For writing there can be many channels for each uuid. Those can
+     * share a websocket (within a specific global), so we have a map
+     * of {uuid: [WebSocket, count]}.  Count is incremented when a
```
Great docs, thanks! In particular noting when the count is decremented is useful.

foolip commented on resources/channel.sub.js:
```diff
@@ -116,16 +132,21 @@
-                "connect", new Set(),
-                "close", new Set()]]);
+            this.eventListeners = {
+                connect:new Set(),
```
```suggestion
                connect: new Set(),
```

foolip commented on resources/channel.sub.js:
```diff
@@ -171,19 +192,17 @@
-            if (!this.eventListeners.has(type)) {
-                this.eventListeners.set(type, new Set());
+            if (!this.eventListeners.hasOwnProperty(type)) {
+                throw new Error(`Unrecognised event type ${type}`);
```
Alright, this settles the questions I had in the previous review about unnecessary (now necessary) code to add empty event handler maps. Thanks!

foolip commented on resources/channel.sub.js:
```diff
@@ -0,0 +1,1003 @@
+        return channel;
+    };
+
+    self.closeAllChannelSockets = async function() {
```
I would say the closest thing to consistent would be to rename this to `close_all_channel_sockets`, since the other global methods added by this file follow that pattern.

foolip commented on resources/channel.sub.js:
```diff
@@ -703,81 +733,119 @@
+     * array) containing a serialized representation of the child
+     * values.
+     *
+     * objectId - An integer if, used to handle object graphs. Where
```
I'm not sure if a more words were intended, but removing a word works:

```suggestion
     * objectId - An integer used to handle object graphs. Where
```

foolip commented on resources/channel.sub.js:
```diff
@@ -703,81 +733,119 @@
 
-        while (stack.length > 0) {
-            const {item, target, targetKey} = stack.shift();
+        /* Instead of making this recursive, use a queue holding the objects to be
```
Yay, great comment!

foolip commented on resources/channel.sub.js:
```diff
@@ -703,81 +733,119 @@
+         * which the serialization of the current item will be added.
+         *
+         * targetName - For serializing object members, the name of
+         * the property. For serialzing maps either "key" or "value",
```
```suggestion
         * the property. For serializing maps either "key" or "value",
```

foolip commented on resources/channel.sub.js:
```diff
@@ -789,74 +857,92 @@
                     break;
                 case "map":
-                    if (targetKey === "key") {
+                    // We always serialize key and value as adjacent items in the queue,
```
Yes!

foolip commented on resources/channel.sub.js:
```diff
@@ -789,74 +857,92 @@
     }
 
+    /**
+     * Deerialize an object from a JSON-compatible representation.
```
```suggestion
     * Deserialize an object from a JSON-compatible representation.
```

foolip commented on websockets/handlers/msg_channel_wsh.py:
```diff
@@ -14,18 +14,47 @@
+# Entries on the queue itself are formed of (command, data) pairs. The
+# command can be either "close", signalling the socket is closing and
+# the reference count on the channel should be decremented, or
+# message, which indicates a message.
```
```suggestion
# "message", which indicates a message.
```

foolip commented on websockets/handlers/msg_channel_wsh.py:
```diff
@@ -140,15 +199,26 @@ def run_write(request, uuid, queue):
-    # Direction of None indicates that we force delete the queue from the stash
+    """Update the channel state in the stash when closing a connection
+
+    This updates the stash entry, inclusing refcounts, once a
```
```suggestion
    This updates the stash entry, including refcounts, once a
```

foolip reviewed (approved):
All that is left is some minor things.

Thanks for getting through this, it must have taken much longer to write than to review :D

jgraham commented:
Thanks for all the reviews @foolip!
